lib/src/http/http_interceptor.dart

```diff
@@ -0,0 +1,73 @@
+library w_transport.src.http.http_interceptor;
+
+import 'dart:async';
+
+import 'package:w_transport/w_transport.dart';
+
+/// Base class representing an interceptor that can be registered with a
+/// [Client] instance to intercept HTTP requests and responses in order to
+/// modify or augment them prior to dispatch and delivery, respectively.
+class HttpInterceptor {
+  Future<RequestPayload> interceptRequest(RequestPayload payload) async {
+    return payload;
+  }
+
+  Future<ResponsePayload> interceptResponse(ResponsePayload payload) async {
+    return payload;
+  }
+}
+
+/// Representation of a request payload. Currently only contains the request
+/// instance, but may contain more contextual information in the future.
+///
+/// The [request] instance can be type checked and cast to [FormRequest],
+/// [JsonRequest], [MultipartRequest], [Request], or [StreamedRequest] as
+/// necessary.
+class RequestPayload {
+  final BaseRequest request;
+  RequestPayload(BaseRequest this.request);
+}
+
+/// Representation of a response payload. Contains the [response] instance, the
+/// finalized [request] that initiated the response, and an [exception] if one
+/// occurred.
+///
+/// The [response] instance can be type checked and cast to [Response] or
+/// [StreamedResponse] as necessary.
+class ResponsePayload {
+  final FinalizedRequest request;
+  BaseResponse response;
+  final RequestException exception;
+  ResponsePayload(FinalizedRequest this.request, BaseResponse this.response,
+      [RequestException this.exception]);
+}
+
+/// Representation of a pathway on which..
+///
+/// 1. payloads of a certain type travel, and
+/// 2. a set of interceptors can be attached.
+///
+/// As a payload travels a pathway, each interceptor (in the order they were
+/// attached) has the opportunity to process, evaluate, modify, augment, or
+/// replace the payload. This interception can be synchronous or asynchronous -
+/// the pathway ensures that each interceptor is applied serially.
+///
+/// Simply put, a pathway accepts a payload and asynchronously returns a
+/// payload.
+class Pathway<T> {
```

Comment:
For now, it's small enough that I'm including it in here and eventually we can break it out.